moveit_core/robot_state/test/robot_state_benchmark.cpp

```diff
@@ -66,312 +61,279 @@ Eigen::Isometry3d createTestIsometry()
-{
-  int n_states = st.range(0);
-  const moveit::core::RobotModelPtr& robot_model = moveit::core::loadTestingRobotModel(PANDA_TEST_ROBOT);
+  std::vector<moveit::core::RobotState> constructStates(size_t num)
+  {
+    std::vector<moveit::core::RobotState> states;
+    states.reserve(num);
+    for (size_t i = 0; i < num; i++)
+      states.emplace_back(robot_model);
+    return states;
+  }
+  std::vector<moveit::core::RobotState> constructStates(size_t num, const moveit::core::RobotState& state)
+  {
+    std::vector<moveit::core::RobotState> states;
+    states.reserve(num);
+    for (size_t i = 0; i < num; i++)
+      states.emplace_back(state);
+    return states;
+  }
 
-  // Make sure the group exists, otherwise exit early with an error.
-  if (!robot_model->hasJointModelGroup(PANDA_TEST_GROUP))
+  std::vector<size_t> randomPermudation(size_t num)
   {
-    st.SkipWithError("The planning group doesn't exist.");
-    return;
+    std::vector<size_t> result;
+    result.reserve(num);
+    for (size_t i = 0; i < num; i++)
+      result.push_back(i);
+
+    std::random_device random_device;
+    std::mt19937 generator(random_device());
+
+    std::shuffle(result.begin(), result.end(), generator);
+    return result;
   }
 
-  // Robot state.
-  moveit::core::RobotState robot_state(robot_model);
-  robot_state.setToDefaultValues();
+  moveit::core::RobotModelPtr robot_model;
+};
 
+// Benchmark time to construct RobotStates
+BENCHMARK_DEFINE_F(RobotStateBenchmark, construct)(benchmark::State& st)
+{
   for (auto _ : st)
   {
-    for (int i = 0; i < n_states; i++)
-    {
-      std::unique_ptr<moveit::core::RobotState> robot_state_copy;
-      benchmark::DoNotOptimize(robot_state_copy = std::make_unique<moveit::core::RobotState>(robot_state));
-      benchmark::ClobberMemory();
-    }
+    auto states = constructStates(st.range(0));
+    benchmark::DoNotOptimize(states);
+    benchmark::ClobberMemory();
   }
 }
 
-// Benchmark time to call `setToRandomPositions` and `update` on a RobotState.
-static void robotStateUpdate(benchmark::State& st)
+// Benchmark time to copy-construct a RobotState.
+BENCHMARK_DEFINE_F(RobotStateBenchmark, copyConstruct)(benchmark::State& st)
 {
-  int n_states = st.range(0);
-  const moveit::core::RobotModelPtr& robot_model = moveit::core::loadTestingRobotModel(PR2_TEST_ROBOT);
   moveit::core::RobotState state(robot_model);
+  state.setToDefaultValues();
+  state.update();
 
   for (auto _ : st)
   {
-    for (int i = 0; i < n_states; ++i)
-    {
-      state.setToRandomPositions();
-      state.update();
-      benchmark::ClobberMemory();
-    }
+    auto states = constructStates(st.range(0), state);
+    benchmark::DoNotOptimize(states);
+    benchmark::ClobberMemory();
   }
 }
 
-// Benchmark time to call `setToRandomPositions` and `getGlobalLinkTransform` on a RobotState.
-static void robotStateForwardKinematics(benchmark::State& st)
+// Benchmark time to call `setToRandomPositions` and `update` on RobotState.
```

Comment:
Should we keep the `getGlobalLinkTransform` version, in addition to `update`?
`update` calls `getGlobalLinkTransform` + `updateCollisionBodyTransforms`, so it does something extra.
Kinematics solvers for instance don't always need the collision body transforms, and use `getGlobalLinkTransform` instead, which should be slightly faster.

Reply:
I consider `update()` simply a real-world test for FK updates: we just want to run many transform updates - no matter being link or collision transforms.

Reply:
That makes sense, but we have code that doesn't care about collision-body transforms, like IK solvers, or Servo (the IK part). They just call `getGlobalLinkTransform`, which updates only the link transforms. I would like to keep that version so that we have a direct measure of how much time that's taking.

Reply:
I think that doesn't make a difference.